node/pkg/watchers/solana/shim.go

```diff
@@ -0,0 +1,362 @@
+) (bool, error) {
+	topLevelIdx := uint16(topLevelIndex)
+	inst := tx.Message.Instructions[topLevelIdx]
+
+	// The only top-level instruction generated by the shim contract is the PostMessage event. Parse that to get
+	// the fields we need to generate an observation.
+	postMessage, err := shimParsePostMessage(s.shimPostMessageDiscriminator, inst.Data)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse top-level shim instruction %d: %w", topLevelIdx, err)
+	}
+
+	if postMessage == nil {
+		return false, nil
+	}
+
+	// Find the set of inner instructions that go with this top level instruction by matching the index.
+	outerIdx := -1
+	for idx, inner := range innerInstructions {
+		if inner.Index == topLevelIdx {
+			outerIdx = int(idx)
+			break
+		}
+	}
+
+	if outerIdx == -1 {
+		return false, fmt.Errorf("failed to find inner instructions for top-level shim instruction %d", topLevelIdx)
+	}
+
+	// Process the inner instructions associated with this shim top-level instruction and produce an observation event.
+	err = s.shimProcessRest(logger, whProgramIndex, shimProgramIndex, tx, innerInstructions[outerIdx].Instructions, outerIdx, 0, postMessage, alreadyProcessed, isReobservation)
+	if err != nil {
+		return false, fmt.Errorf("failed to process inner instructions for top-level shim instruction %d: %w", topLevelIdx, err)
+	}
+
+	return true, nil
+}
+
+// shimProcessInnerInstruction handles an inner instruction where the program ID matches the shim contract. It does the following:
+// - Verifies that the instruction is a shim PostMessage. If not, it just returns. If it is, it parses it.
+// - Searches through the subsequent inner instructions in the set to find the entry for the core. Makes sure it is unreliable with no payload.
+// - Searches for the inner shim MessageEvent to get the remaining fields needed to generate the observation.
+// - Publishes the observation.
+func (s *SolanaWatcher) shimProcessInnerInstruction(
+	logger *zap.Logger,
+	whProgramIndex uint16,
+	shimProgramIndex uint16,
+	tx *solana.Transaction,
+	innerInstructions []solana.CompiledInstruction,
+	outerIdx int,
+	startIdx int,
+	alreadyProcessed ShimAlreadyProcessed,
+	isReobservation bool,
+) (bool, error) {
+	// See if this is a PostMessage event from the shim contract. If so, parse it. If not, bail out now.
+	postMessage, err := shimParsePostMessage(s.shimPostMessageDiscriminator, innerInstructions[startIdx].Data)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse inner shim post message instruction %d, %d: %w", outerIdx, startIdx, err)
+	}
+
+	if postMessage == nil {
+		return false, nil
+	}
+
+	alreadyProcessed.add(outerIdx, startIdx)
+
+	err = s.shimProcessRest(logger, whProgramIndex, shimProgramIndex, tx, innerInstructions, outerIdx, startIdx+1, postMessage, alreadyProcessed, isReobservation)
+	if err != nil {
+		return false, fmt.Errorf("failed to process inner instructions for top-level shim instruction %d: %w", outerIdx, err)
```

Comment:
this is failing to process the inner instruction for an inner shim post message instruction, right?